File: 2021/10/09/main.py

class TrieNode():
    def __init__(self):
        self.next = {}
        self.word = ""

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        r, n, m = [], len(board[0]), len(board)
        
        trie = TrieNode()
        for word in words:
            cur = trie
            for c in word:
                if c not in cur.next:
                    cur.next[c] = TrieNode()
                cur = cur.next[c]
            cur.word = word

        def dfs(i, j, trie):
            c = board[i][j]
            if c == '#' or c not in trie.next:
                return

            trie = trie.next[c]
            if len(trie.word) > 0:
                r.append(trie.word)
                trie.word = ""

            board[i][j] = '#'
            if i > 0:
                dfs(i-1, j, trie)

            if i < len(board)-1:
                dfs(i+1, j, trie)

            if j > 0:
                dfs(i, j-1, trie)

            if j < len(board[0])-1:
                dfs(i, j+1, trie)

            board[i][j] = c

        for i in range(m):
            for j in range(n):
                dfs(i, j, trie)

        return r

File: 2021/10/09/main.go
package main

type TrieNode struct {
	next [26]*TrieNode
	word string
}

func findWords(board [][]byte, words []string) []string {
	var r []string

	m := len(board)
	if m == 0 {
		return r
	}

	n := len(board[0])
	if n == 0 {
		return r
	}

	trie := buildTrie(words)

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			dfs(board, i, j, trie, &r)
		}
	}

	return r
}

func dfs(board [][]byte, i int, j int, trie *TrieNode, r *[]string) {
	c := board[i][j]
	if c == '#' || trie.next[int(c-'a')] == nil {
		return
	}

	trie = trie.next[int(c-'a')]
	if len(trie.word) > 0 {
		// Found one
		*r = append(*r, trie.word)
		trie.word = ""
	}

	board[i][j] = '#'
	if i > 0 {
		dfs(board, i-1, j, trie, r)
	}

	if i < len(board)-1 {
		dfs(board, i+1, j, trie, r)
	}

	if j > 0 {
		dfs(board, i, j-1, trie, r)
	}

	if j < len(board[0])-1 {
		dfs(board, i, j+1, trie, r)
	}

	board[i][j] = c
}

func buildTrie(words []string) *TrieNode {
	root := new(TrieNode)
	for _, word := range words {
		cur := root
		for _, c := range word {
			cidx := int(c - 'a')
			if cur.next[cidx] == nil {
				cur.next[cidx] = new(TrieNode)
			}
			cur = cur.next[cidx]
		}
		cur.word = word
	}

	return root
}
